rust/src/container.rs

```diff
@@ -397,30 +398,16 @@ pub fn container_encapsulate(args: Vec<String>) -> CxxResult<()> {
         })
         .collect::<Result<_>>()?;
 
-    let handle = tokio::runtime::Handle::current();
-    let package_structure = match &opt.prior_build {
-        Some(prior_build) => {
-            let prev_package_structure = handle.block_on(async {
-                let proxy = containers_image_proxy::ImageProxy::new().await?;
-                let mut annotation_build: Vec<Vec<String>> = Vec::new();
-                let oi = proxy.open_image(prior_build).await?;
-                let (_digest, manifest) = proxy.fetch_manifest(&oi).await?;
-                let layers = manifest.layers();
-                for layer in layers {
-                    let annotation_layer = layer
-                        .annotations()
-                        .as_ref()
-                        .expect("Layer does not consist annotation");
-                    let pkgs: Vec<&str> = annotation_layer["Content"].split(',').collect();
-                    let mut pkgs_string: Vec<String> = Vec::new();
-                    for pkg in &pkgs {
-                        pkgs_string.push(pkg.to_string());
-                    }
-                    annotation_build.push(pkgs_string);
-                }
-                Ok::<_, anyhow::Error>(annotation_build)
-            })?;
-            Some(prev_package_structure)
+    let package_structure = match &opt.ocimeta_json {
+        Some(ocimeta_json) => {
+            let path = std::path::Path::new(ocimeta_json);
+            if path.is_file() {
+                let manifest = oci_spec::image::ImageManifest::from_file(path)
+                    .expect("parse ImageManifest from ocimeta");
+                Some(manifest)
```

Comment:
Due to the change in `ostree_ext::container::encapsulate` args for `prior-build-metadata` from `Result<oci_spec::Image::Manifest>` to `Result<&oci_spec::Image::Manifest>`. How would this value be returned without facing error?

Reply:
Your currently pushed code builds for me with this change:
```
diff --git a/rust/src/container.rs b/rust/src/container.rs
index 92e3c44b..b5a4ae8e 100644
--- a/rust/src/container.rs
+++ b/rust/src/container.rs
@@ -434,7 +434,7 @@ pub fn container_encapsulate(args: Vec<String>) -> CxxResult<()> {
                 repo,
                 rev.as_str(),
                 &config,
-                package_structure,
+                package_structure.as_ref(),
                 Some(opts),
                 Some(meta),
                 &opt.imgref,
```

Reply:
In general given a T, one can always convert that into a &T.  An interface that accepts a T "takes ownership" of the value so the caller can't use it again.  Hence APIs (functions) that accept a &T are more flexible for the caller.

In Rust it's very common to have Option<&T> (not &Option<T>), and it's the `as_ref()` API which goes from `Option<T>` to `Option<&T>`.

Reply:
ok thanks!